File: src/test/kotlin/school/charset/app/domain/CodePointTest.kt
package school.charset.app.domain

import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.FreeSpec
import io.kotest.matchers.shouldBe

class CodePointTest :
    FreeSpec({

        "init" - {
            "rejects negative values" { shouldThrow<IllegalArgumentException> { CodePoint(-1) } }
            "rejects values above U+10FFFF" { shouldThrow<IllegalArgumentException> { CodePoint(0x110000) } }
        }

        "isBmp" - {
            "true for U+0000 (low boundary)" { CodePoint(0x0000).isBmp shouldBe true }
            "true for U+D800 (surrogates are in the BMP)" { CodePoint(0xD800).isBmp shouldBe true }
            "true for U+FFFF (high boundary)" { CodePoint(0xFFFF).isBmp shouldBe true }
            "false for U+10000 (first code point above the BMP)" { CodePoint(0x10000).isBmp shouldBe false }
            "false for U+10FFFF (max code point)" { CodePoint(0x10FFFF).isBmp shouldBe false }
        }

        "isSurrogate" - {
            "false for U+D7FF (just before the range)" { CodePoint(0xD7FF).isSurrogate shouldBe false }
            "true for U+D800 (range start)" { CodePoint(0xD800).isSurrogate shouldBe true }
            "true for U+DFFF (range end)" { CodePoint(0xDFFF).isSurrogate shouldBe true }
            "false for U+E000 (just after the range)" { CodePoint(0xE000).isSurrogate shouldBe false }
        }
    })

File: test/charset/encoding/code_point_test.exs
defmodule Charset.Encoding.CodePointTest do
  use ExUnit.Case, async: true

  import Charset.Encoding.CodePoint

  doctest Charset.Encoding.CodePoint

  describe "is_code_point/1" do
    test "rejects negative values", do: refute(is_code_point(-1))
    test "rejects values above U+10FFFF", do: refute(is_code_point(0x110000))
    test "rejects non-integers", do: refute(is_code_point("A"))
    test "accepts U+0000 (low boundary)", do: assert(is_code_point(0x0000))
    test "accepts U+10FFFF (high boundary)", do: assert(is_code_point(0x10FFFF))
  end

  describe "is_bmp/1" do
    test "true for U+0000 (low boundary)", do: assert(is_bmp(0x0000))
    test "true for U+D800 (surrogates are in the BMP)", do: assert(is_bmp(0xD800))
    test "true for U+FFFF (high boundary)", do: assert(is_bmp(0xFFFF))
    test "false for U+10000 (first code point above the BMP)", do: refute(is_bmp(0x10000))
    test "false for U+10FFFF (max code point)", do: refute(is_bmp(0x10FFFF))
  end

  describe "is_surrogate/1" do
    test "false for U+D7FF (just before the range)", do: refute(is_surrogate(0xD7FF))
    test "true for U+D800 (range start)", do: assert(is_surrogate(0xD800))
    test "true for U+DFFF (range end)", do: assert(is_surrogate(0xDFFF))
    test "false for U+E000 (just after the range)", do: refute(is_surrogate(0xE000))
  end
end
